### PR TITLE
[ConstraintSystem] More changes to key path inference

### PR DESCRIPTION
**Pull request by xedin**
- Infer key path root bindings transitively

  As a first step in delaying key path type until all of the members are
  resolved, attempt to infer a type of root based on bindings associated
  with key path type.

-  Delay key path type inference until literal capability is known

   Since key path root is now transitively inferred. Key path type
   inference can be delayed until key path is resolved enough to
   infer its capability.

   This solves multiple problems:

   - Inference fully controls what key path type is bound to;
   - KeyPath constraint simplification doesn't have to double-check
      the capability and attempt to re-bind key path type;
   - Custom logic to resolve key path type is no longer necessary;
   - Diagnostics are improved because capability and root/value type
      mismatch are diagnosed when key path is matched against the
      contextual type.

- Improve contextual mismatch diagnostics related to key path literals.

### CODE REVIEW

**Comment by xedin** [comment]
@swift-ci please test

**Comment by xedin** [comment]
@swift-ci please test source compatibility

**Comment by xedin** [comment]
The main idea here is that the solver gets to infer the root transitively from the bindings associated with key path type, that allows all of the references in a key path literal to be resolved and subsequently the key path type can be determined based on the capability instead of eagerly binding it to a contextual type, this give the solver more control over key path inference and helps with diagnostics because fixes could be applied at the point where key path type is converted a contextual type (i.e. parameter type or an explicit initializer type).

File: lib/Sema/CSBindings.cpp [hunk 1]
```diff
@@ -538,6 +591,90 @@ void BindingSet::finalize(
...
+        SmallSetVector<PotentialBinding, 4> updatedBindings;
+        for (const auto &binding : Bindings) {
+          auto bindingTy = binding.BindingType->lookThroughAllOptionalTypes();
+
+          assert(isKnownKeyPathType(bindingTy) ||
+                 bindingTy->is<FunctionType>());
+
+          // Functions don't have capability so we can simply add them.
+          if (bindingTy->is<FunctionType>())
+            updatedBindings.insert(binding);
```
Comment (xedin) [inline lib/Sema/CSBindings.cpp:654]: I think in this case we can form a function type that uses root/value type variables and infers parameter flags from the binding, does that make sense, @Jumhyn? If we can make it work then I can remove remaining root/value matching logic from `simplifyKeyPathConstraint`.

Comment (xedin) [reply lib/Sema/CSBindings.cpp:654]: Actually I think I can approach that from a different angle. I Remembered that there is a hack to eagerly bind key path type if the contextual type is a function type. I think once I remove that everything and replace function result type with key path's value type everything is going to fall into place.

**Comment by xedin** [comment]
@swift-ci please test

**Comment by xedin** [comment]
@swift-ci please test source compatibility

File: lib/Sema/CSBindings.cpp (hunk 1 above)
Comment (Jumhyn) [reply lib/Sema/CSBindings.cpp]: Sorry @xedin was out of town last week, but the current solution LGTM!

**Pull request merged**